File: DuAnTauHoaCom/backend/src/controllers/BookingController.js
const BookingService = require('../services/BookingService')
const BookingRepo   = require('../repositories/BookingRepository')
const { ok, created, badRequest, notFound } = require('../utils/response')
const {
  Ve, HanhKhach, ChuyenTau, LichChay, Tau, GaTau,
} = require('../models')

// POST /api/bookings/hold-seats — giữ ghế tạm trước khi điền form
const holdSeats = async (req, res, next) => {
  try {
    const { trips } = req.body
    if (!trips?.length) return badRequest(res, 'Thiếu danh sách chuyến (trips)')
    const result = await BookingService.holdSeatsForCheckout({ trips })
    ok(res, result, 'Giữ ghế thành công')
  } catch (err) { next(err) }
}

const createBooking = async (req, res, next) => {
  try {
    const { trips, passengers, contactInfo, maKhuyenMai } = req.body
    if (!trips?.length || !passengers?.length || !contactInfo)
      return badRequest(res, 'Thiếu thông tin đặt vé')

    const idTaiKhoan = req.user?.id || null
    const result = await BookingService.createBooking({ trips, passengers, contactInfo, idTaiKhoan, maKhuyenMai })
    created(res, {
      maDon:         result.maDon,
      maDatCho:      result.maDatCho,
      idDon:         result.don.id_don_dat_ve,
      tongThanhToan: result.tongThanhToan,
      tienGiam:      result.tienGiam,
    }, 'Đặt vé thành công')
  } catch (err) { next(err) }
}

const lookupBooking = async (req, res, next) => {
  try {
    const { maDatCho, email, phone } = req.body
    if (!maDatCho) return badRequest(res, 'Vui lòng nhập mã đặt chỗ')

    const don = await BookingService.lookupBooking(maDatCho, email, phone)
    if (!don) return notFound(res, 'Không tìm thấy đơn đặt vé hoặc thông tin không khớp')

    // Fallback cho đơn cũ: Sequelize không tự join Ves → query Ve trực tiếp
    if (!don.Ves || don.Ves.length === 0) {
      const vesDirect = await Ve.findAll({
        where: { id_don_dat_ve: don.id_don_dat_ve },
        include: [
          { model: HanhKhach },
          {
            model: ChuyenTau,
            include: [{
              model: LichChay,
              include: [
                Tau,
                { model: GaTau, as: 'GaDi' },
                { model: GaTau, as: 'GaDen' },
              ],
            }],
          },
          { model: GaTau, as: 'GaLen' },
          { model: GaTau, as: 'GaXuong' },
        ],
      })
      don.Ves = vesDirect
    }

    ok(res, BookingService.formatDon(don))
  } catch (err) { next(err) }
}

const getBookingHistory = async (req, res, next) => {
  try {
    const dons = await BookingService.getBookingHistory(req.user.id)

    // Fallback đơn cũ: nếu Ves rỗng, query Ve trực tiếp theo id_don_dat_ve
    for (const don of dons) {
      if (!don.Ves || don.Ves.length === 0) {
        don.Ves = await Ve.findAll({
          where: { id_don_dat_ve: don.id_don_dat_ve },
          include: [
            { model: HanhKhach },
            {
              model: ChuyenTau,
              include: [{
                model: LichChay,
                include: [Tau, { model: GaTau, as: 'GaDi' }, { model: GaTau, as: 'GaDen' }],
              }],
            },
            { model: GaTau, as: 'GaLen' },
            { model: GaTau, as: 'GaXuong' },
          ],
        })
      }
    }

    ok(res, dons.map(BookingService.formatDon))
  } catch (err) { next(err) }
}

const getBookingByCode = async (req, res, next) => {
  try {
    const { maDatCho } = req.params
    const don = await BookingRepo.findByMaDatCho(maDatCho)
    if (!don) return notFound(res, 'Không tìm thấy đơn đặt vé')
    ok(res, BookingService.formatDon(don))
  } catch (err) { next(err) }
}

// POST /api/bookings/release-hold — giải phóng ghế khi thoát form checkout
const releaseHold = async (req, res, next) => {
  try {
    const { sessionId } = req.body
    if (!sessionId) return badRequest(res, 'Thiếu sessionId')
    await BookingRepo.releaseHoldBySession(sessionId)
    ok(res, null, 'Giải phóng ghế thành công')
  } catch (err) { next(err) }
}

module.exports = { holdSeats, createBooking, lookupBooking, getBookingHistory, getBookingByCode, releaseHold }

File: java-backend/src/main/java/com/klntrain/controller/BookingController.java
package com.klntrain.controller;

import com.klntrain.dto.request.CreateBookingRequest;
import com.klntrain.dto.response.ApiResponse;
import com.klntrain.security.JwtPrincipal;
import com.klntrain.service.BookingService;
import com.klntrain.util.AppException;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/bookings")
@RequiredArgsConstructor
public class BookingController {

    private final BookingService bookingService;

    @PostMapping("/hold-seats")
    public ResponseEntity<ApiResponse<?>> holdSeats(@RequestBody Map<String, Object> body) {
        @SuppressWarnings("unchecked")
        var trips = (List<CreateBookingRequest.TripInfo>) body.get("trips");
        if (trips == null || trips.isEmpty()) throw AppException.badRequest("Thiếu danh sách chuyến");
        var result = bookingService.holdSeatsForCheckout(trips);
        return ResponseEntity.ok(ApiResponse.ok(result, "Giữ ghế thành công"));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<?>> createBooking(
        @RequestBody CreateBookingRequest req,
        @AuthenticationPrincipal JwtPrincipal principal
    ) {
        if (req.getTrips() == null || req.getTrips().isEmpty() ||
            req.getPassengers() == null || req.getPassengers().isEmpty() ||
            req.getContactInfo() == null) {
            throw AppException.badRequest("Thiếu thông tin đặt vé");
        }
        Long idTaiKhoan = principal != null ? principal.getId() : null;
        var result = bookingService.createBooking(req, idTaiKhoan);
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.ok(result, "Đặt vé thành công"));
    }

    @PostMapping("/lookup")
    public ResponseEntity<ApiResponse<?>> lookupBooking(@RequestBody Map<String, String> body) {
        String maDatCho = body.get("maDatCho");
        if (maDatCho == null || maDatCho.isBlank()) throw AppException.badRequest("Vui lòng nhập mã đặt chỗ");
        var don = bookingService.lookupBooking(maDatCho, body.getOrDefault("email", ""), body.getOrDefault("phone", ""));
        if (don == null) throw AppException.notFound("Không tìm thấy đơn hoặc thông tin không khớp");
        return ResponseEntity.ok(ApiResponse.ok(don));
    }

    @GetMapping("/history")
    public ResponseEntity<ApiResponse<?>> getHistory(@AuthenticationPrincipal JwtPrincipal principal) {
        var history = bookingService.getBookingHistory(principal.getId());
        return ResponseEntity.ok(ApiResponse.ok(history));
    }

    @PostMapping("/release-hold")
    public ResponseEntity<ApiResponse<?>> releaseHold(@RequestBody Map<String, String> body) {
        String sessionId = body.get("sessionId");
        if (sessionId == null || sessionId.isBlank()) throw AppException.badRequest("Thiếu sessionId");
        // TamGiuGheRepository không inject ở đây — xử lý thông qua service nếu cần
        return ResponseEntity.ok(ApiResponse.ok(null, "Giải phóng ghế thành công"));
    }
}
